File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/datavalidation/pogo/EmailControllerSpec.kt
package io.micronaut.docs.datavalidation.pogo

import io.kotest.matchers.shouldBe
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.micronaut.context.ApplicationContext
import io.micronaut.http.HttpRequest
import io.micronaut.http.HttpStatus
import io.micronaut.http.client.HttpClient
import io.micronaut.http.client.exceptions.HttpClientResponseException
import io.micronaut.runtime.server.EmbeddedServer

class EmailControllerSpec: StringSpec() {

    val embeddedServer = autoClose(
        ApplicationContext.run(EmbeddedServer::class.java, mapOf("spec.name" to "datavalidationpogo"))
    )

    val client = autoClose(
        embeddedServer.applicationContext.createBean(HttpClient::class.java, embeddedServer.url)
    )

    init {
        //tag::pojovalidated[]
        "test pojo validation" {
            val e = shouldThrow<HttpClientResponseException> {
                val email = Email()
                email.subject = "Hi"
                email.recipient = ""
                client.toBlocking().exchange<Email, Any>(HttpRequest.POST("/email/send", email))
            }
            var response = e.response

            response.status shouldBe HttpStatus.BAD_REQUEST

            val email = Email()
            email.subject = "Hi"
            email.recipient = "me@micronaut.example"
            response = client.toBlocking().exchange<Email, Any>(HttpRequest.POST("/email/send", email))

            response.status shouldBe HttpStatus.OK
        }
        //end::pojovalidated[]
    }
}

File: test-suite-python/src/test/python/micronaut/docs/datavalidation/pogo/EmailControllerSpec.py
from typing import Annotated

import java
from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.http.client.exceptions import HttpClientResponseException
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Disabled, Test

HttpRequest = java.type("io.micronaut.http.HttpRequest")
HttpStatus = java.type("io.micronaut.http.HttpStatus")
Email = java.type("micronaut.docs.datavalidation.pogo.Email")


@Property(name="spec.name", value="datavalidationpogo")
@MicronautTest
class EmailControllerSpec:
    client: Annotated[HttpClient, Inject, Client("/")]

    # tag::pojovalidated[]
    @Test
    @Disabled("GraalPy Java exception matching currently fails for the propagated body validation exception")
    def test_pojo_validation(self):
        try:
            email = Email()
            email.setSubject("Hi")
            email.setRecipient("")
            self.client.toBlocking().exchange(HttpRequest.POST("/email/send", email))
            assert False
        except HttpClientResponseException as e:
            response = e.getResponse()

        assert response.getStatus() == HttpStatus.BAD_REQUEST

        email = Email()
        email.setSubject("Hi")
        email.setRecipient("me@micronaut.example")
        response = self.client.toBlocking().exchange(HttpRequest.POST("/email/send", email))

        assert response.getStatus() == HttpStatus.OK
    # end::pojovalidated[]
